src/wp-admin/includes/theme-previews.php

```diff
@@ -0,0 +1,129 @@
+<?php
+/**
+ * Enable theme previews in the Site Editor for block themes.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Filters the blog option to return the path for the previewed theme.
+ *
+ * @param string $current_stylesheet The current theme's stylesheet or template path.
+ * @return string The previewed theme's stylesheet or template path.
+ */
+function gutenberg_get_theme_preview_path( $current_stylesheet = null ) {
+	// Don't allow non-admins to preview themes.
+	if ( ! current_user_can( 'switch_themes' ) ) {
+		return $current_stylesheet;
+	}
+
+	$preview_stylesheet = ! empty( $_GET['theme_preview'] ) ? $_GET['theme_preview'] : null;
+	$wp_theme           = wp_get_theme( $preview_stylesheet );
+	if ( ! is_wp_error( $wp_theme->errors() ) ) {
+		if ( current_filter() === 'template' ) {
+			$theme_path = $wp_theme->get_template();
+		} else {
+			$theme_path = $wp_theme->get_stylesheet();
+		}
+
+		return sanitize_text_field( $theme_path );
+	}
+
+	return $current_stylesheet;
+}
+
+/**
+ * Adds a middleware to the REST API to set the theme for the preview.
+ */
+function gutenberg_attach_theme_preview_middleware() {
+	// Don't allow non-admins to preview themes.
+	if ( ! current_user_can( 'switch_themes' ) ) {
+		return;
+	}
+	
+	wp_add_inline_script(
+		'wp-api-fetch',
+		sprintf(
+			'wp.apiFetch.use( wp.apiFetch.createThemePreviewMiddleware( %s ) );',
+			wp_json_encode( sanitize_text_field( $_GET['theme_preview'] ) )
+		),
+		'after'
+	);
+}
+
+/**
+ * Temporary function to add a live preview button to block themes.
```

Comment:
This one should be ported to modify the themes showcase file to include the button and all this jquery script should be gone.